pyqtorch/utils.py

```diff
@@ -114,5 +115,48 @@ def _rand(n_qubits: int) -> torch.Tensor:
     return _state.reshape([2] * n_qubits + [batch_size]).to(device=device)
 
 
-def param_dict(keys: Sequence[str], values: Sequence[torch.Tensor]) -> dict[str, torch.Tensor]:
+def param_dict(keys: Sequence[str], values: Sequence[Tensor]) -> dict[str, Tensor]:
     return {key: val for key, val in zip(keys, values)}
+
+
+def density_mat(state: Tensor) -> Tensor:
```

Comment:
```
def density_mat(state: Tensor) -> Tensor:
    """
    Computes the density matrix from a pure state vector.

    Args:
        state (Tensor): The pure state vector :math:`|\\psi\\rangle`.

    Returns:
        Tensor: The density matrix :math:`\\rho = |\psi \\rangle \\langle\\psi|`.
    """
    n_qubits = len(state.size()) - 1
    batch_dim = state.dim() - 1
    batch_size = state.shape[-1]
    batch_first_perm = [batch_dim] + list(range(batch_dim))
    state = torch.permute(state, batch_first_perm).reshape(batch_size, 2**n_qubits)
    undo_perm = (1, 2, 0)
    return torch.permute(torch.einsum('bi,bj->bij', (state, state.conj())), undo_perm)
```

Reply:
@Roland-djee @EthanObadia lets use this version, it avoids the python looping and instead does one einsum operation